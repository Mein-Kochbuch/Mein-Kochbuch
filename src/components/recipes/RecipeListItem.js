import React from "react";
import styled from "styled-components/native";
import {Colors} from "react-native/Libraries/NewAppScreen";
import {useColorScheme} from "react-native-appearance";

export default function RecipeListItem({recipe}) {
    const isDarkMode = useColorScheme() === 'dark';

    const imageSource = recipe.thumbnail_url ? {uri: recipe.thumbnail_url} : require("../../../resources/platzhalter.png")

    return (
        <StyledTouchableOpacity isDarkMode={isDarkMode}>
            <ImageStyled source={imageSource}/>
            <StyledTest numberOfLines={2} ellipsizeMode={"middle"}>
                {recipe.title}
            </StyledTest>
        </StyledTouchableOpacity>
    )
}

const ImageStyled = styled.Image`
  height: 100px;
  width: 100px;
  padding-top: 20px;
  padding-bottom: 20px;
  background-color: #68a0cf;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`

const StyledTest = styled.Text`
  text-align: center;
  font-size: 24px;
  margin-left: 12px;
  max-width: 70%;
`

const StyledTouchableOpacity = styled.TouchableOpacity`
  padding: 12px;
  margin: 12px;
  height: 120px;
  display: flex;
  flex-direction: row;
  align-items: center;
  border-style: solid;
  border-color: ${props => props.isDarkMode ? Colors.lighter : Colors.darker};
  border-radius: 10px;
  border-width: 2px;
`
