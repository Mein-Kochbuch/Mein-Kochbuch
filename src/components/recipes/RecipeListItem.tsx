import React from 'react';
import styled from 'styled-components/native';
import {Colors} from 'react-native/Libraries/NewAppScreen';
import {Appearance, ListRenderItemInfo} from 'react-native';
import {useHistory} from 'react-router-native';
import {RecipePreview} from '../../models/RecipePreview';

export type RecipeListItemProps = {
  listItem: ListRenderItemInfo<RecipePreview>;
};

export default function RecipeListItem({listItem}: RecipeListItemProps) {
  const history = useHistory();
  const isDarkMode = Appearance.getColorScheme() === 'dark';

  const recipe = listItem.item;
  const imageSource = recipe.thumbnail?.url
    ? {uri: recipe.thumbnail.url}
    : require('../../../resources/platzhalter.png');

  const handleOnPress = () => {
    history.push('/recipes/' + recipe.id);
  };

  return (
    <StyledTouchableOpacity isDarkMode={isDarkMode} onPress={handleOnPress}>
      <ImageStyled source={imageSource} />
      <StyledTest numberOfLines={2} ellipsizeMode={'middle'}>
        {recipe.title}
      </StyledTest>
    </StyledTouchableOpacity>
  );
}

const ImageStyled = styled.Image`
  height: 100px;
  width: 100px;
  padding-top: 20px;
  padding-bottom: 20px;
  background-color: #68a0cf;
  border-radius: 10px;
  border-width: 1px;
  border-color: #fff;
`;

const StyledTest = styled.Text`
  text-align: center;
  font-size: 24px;
  margin-left: 12px;
  max-width: 70%;
`;

interface StyledTouchableOpacityProps {
  isDarkMode: boolean;
}

const StyledTouchableOpacity = styled.TouchableOpacity<StyledTouchableOpacityProps>`
  padding: 12px;
  margin: 12px;
  height: 120px;
  display: flex;
  flex-direction: row;
  align-items: center;
  border-style: solid;
  border-color: ${props => (props.isDarkMode ? Colors.lighter : Colors.darker)};
  border-radius: 10px;
  border-width: 2px;
`;
