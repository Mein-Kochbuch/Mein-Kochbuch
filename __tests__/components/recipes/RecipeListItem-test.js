import 'react-native';
import React from 'react';

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import {Image} from 'react-native';
import RecipeListItem from '../../../src/components/recipes/RecipeListItem';
import {fireEvent, render} from '@testing-library/react-native';
import {Router} from 'react-router-native';

describe('RecipeListItem Test', () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('light', () => {
    const item = {
      item: {
        title: 'test-title',
      },
    };

    const component = renderer.create(<RecipeListItem item={item} />);

    expect(component.toJSON()).toMatchSnapshot();
    expect(component.toJSON().children.length).toBe(2);
    expect(component.toJSON().children[1].children[0]).toBe('test-title');

    const testInstance = component.root;

    expect(
      testInstance.findAllByType(RecipeListItem)[0].props.item.item,
    ).toStrictEqual({title: 'test-title'});
    expect(testInstance.findAllByType(Image)[0].props.source).toStrictEqual({
      testUri: '../../../resources/platzhalter.png',
    });
  });

  it('click', () => {
    const history = {
      push: jest.fn(),
      listen: jest.fn(),
      location: {pathname: '/'},
    };

    const item = {
      item: {
        pk: '3',
        title: 'test-title',
      },
    };

    const {getByText} = render(
      <Router history={history}>
        <RecipeListItem item={item} />
      </Router>,
    );

    fireEvent(getByText('test-title'), 'onPress');
    expect(history.push).toBeCalledWith('/recipes/3');
  });
});
