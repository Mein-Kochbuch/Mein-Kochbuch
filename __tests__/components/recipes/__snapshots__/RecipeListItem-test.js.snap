// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeListItem Test light 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "borderColor": "#222",
      "borderRadius": 10,
      "borderStyle": "solid",
      "borderWidth": 2,
      "display": "flex",
      "flexDirection": "row",
      "height": 120,
      "marginBottom": 12,
      "marginLeft": 12,
      "marginRight": 12,
      "marginTop": 12,
      "opacity": 1,
      "paddingBottom": 12,
      "paddingLeft": 12,
      "paddingRight": 12,
      "paddingTop": 12,
    }
  }
>
  <Image
    source={
      Object {
        "testUri": "../../../resources/platzhalter.png",
      }
    }
    style={
      Array [
        Object {
          "backgroundColor": "#68a0cf",
          "borderColor": "#fff",
          "borderRadius": 10,
          "borderWidth": 1,
          "height": 100,
          "paddingBottom": 20,
          "paddingTop": 20,
          "width": 100,
        },
      ]
    }
  />
  <Text
    ellipsizeMode="middle"
    numberOfLines={2}
    style={
      Array [
        Object {
          "fontSize": 24,
          "marginLeft": 12,
          "maxWidth": "70%",
          "textAlign": "center",
        },
      ]
    }
  >
    test-title
  </Text>
</View>
`;
